### PR TITLE
Add support for reading TIFFs with PlanarConfiguration=2

### LINKED ISSUE #4621: Corruption of TIFF images that was introduced in 5.0.0
I’m one of the maintainers of Loris, an image server that uses Pillow. One of our users has reported corruption of TIFF images that I believe comes from Pillow (https://github.com/loris-imageserver/loris/issues/503).

This example downloads the TIFF, and converts it to a JPEG (image from https://elifesciences.org/digests/55692/could-some-antibiotics-be-immune-stimulants):

```python
import hashlib
import os
import shutil
from urllib.request import urlretrieve

from PIL import Image

# Download the image
filename, _ = urlretrieve("https://prod-elife-published.s3.amazonaws.com/digests/55692/digest-55692.tif")

# Expected size: 219528
# Expected hash: 01ddb8885aa5252b41be06dcdf36703932be24436013d5ea65ca9b1fd675258a
print(f"size = {os.stat(filename).st_size}")
print(f"sha256 = {hashlib.sha256(open(filename, 'rb').read()).hexdigest()}")

# Open the image as a TIF; save as a JPEG
im = Image.open(filename)
im.save("digest-55692.jpg")
```

This table shows the output from three versions of Pillow (resized for readability):

<table>
<tr>
<td></td>
<td></td>
<td></td>
</tr><tr><td>4.3.0</td><td>5.0.0 (next version)</td><td>7.1.2 (latest)
</td></tr></table>

The corruption is not deterministic; running the script repeatedly gets images that look different every time.

I have attached [a zip archive (`pillow_test_cases.zip`)](https://github.com/python-pillow/Pillow/files/4616394/pillow_test_cases.zip) which includes the original TIFF, and the three JPEG outputs.

I produced these examples running inside a Docker container, built with the following Dockerfile template and varying the version of Pillow:

```dockerfile
FROM ubuntu:18.04

RUN apt-get update
RUN apt-get install -y python3 python3-pip

RUN pip3 install Pillow==5.0.0

COPY example.py /

CMD ["python3", "/example.py"]
```

We've had issues with TIFFs in Loris/Pillow in the past, which led to https://github.com/python-pillow/Pillow/issues/2926. I wonder if this issue has a similar cause?

### PR DESCRIPTION
**Pull request by kkopachev**
Fixes #4621 .

Adding support for reading of compressed TIFFs with planar config = separate.

Since TIFF reading is done in tiles or strips (might be multiple rows), I've added reading it by sample and shuffling into image data. For 16bit images I had to create new single channel unpackers.

There is also special handling for RGBa images with separate plane storage, since unpacking it sample by sample would leave data in RGBa, so there is extra step required to convert it to RGBA.

I also thought of using line-interleaved unpackers, but they seem to be no use since tiles smaller than width of image.

### CODE REVIEW

**Comment by kkopachev** [comment]
Thanks @radarhere for lint fixes. 

Not sure what's going on with MacOS builds and why they are failing. I don't have Mac in possession to test it first hand, so I tried enabling `TRACE` logs for TiffDecode [and pushed them](https://github.com/kkopachev/Pillow/commit/e13704b9cb6e47afac68a71887a2bd0e9792358e) into separate branch under my form to see in GA where it crashes. Surprisingly it [finished with no errors](https://github.com/kkopachev/Pillow/runs/702990933). So I don't have any ideas where to look. 
Any help appreciated.

File: src/libImaging/TiffDecode.c [hunk 1]
```diff
@@ -384,29 +435,31 @@ int ImagingLibTiffDecode(Imaging im, ImagingCodecState state, UINT8* buffer, Py_
...
+                    current_tile_width = min(tile_width, state->xsize - x);
+
+                    // iterate over each line in the tile and stuff data into image
+                    for (tile_y = 0; tile_y < min(tile_length, state->ysize - y); tile_y++) {
+                        TRACE(("Writing tile data at %dx%d using tile_width: %d; \n", tile_y + y, x, current_tile_width));
+
+                        // UINT8 * bbb = state->buffer + tile_y * row_byte_size;
+                        // TRACE(("chars: %x%x%x%x\n", ((UINT8 *)bbb)[0], ((UINT8 *)bbb)[1], ((UINT8 *)bbb)[2], ((UINT8 *)bbb)[3]));
+
+                        ((*unpacker)[plane])((UINT8*) im->image[tile_y + y] + x * im->pixelsize,
```
Comment (radarhere) [inline src/libImaging/TiffDecode.c:458]: I'm having any luck yet in coming up with a fix, but the `unpacker` pointer here is a starting point for why the macOS tests are failing.

**Comment by radarhere** [comment]
I've created PR https://github.com/kkopachev/Pillow/pull/2 to resolve the macOS exception

**Comment by kkopachev** [comment]
@radarhere Thank you for update, looks cleaner. I was trying to use higher-level language approach to swap pointer :) 
I'm curious why original solution did not work. Any insights why it was failing?

**Comment by radarhere** [comment]
No insights. If it was a simple mistake, then you would think it would be failing for more platforms than just macOS. I just found that `((*unpacker)[plane])` was where the error was being triggered, so I started trying alternatives and somewhere in there thought to remove the need for dereferencing entirely. So some of it was just trial and error using my macOS machine, some of it was using trial and error on GitHub Actions.

File: src/libImaging/Unpack.c [hunk 2]
```diff
@@ -1263,6 +1263,94 @@ band3I(UINT8* out, const UINT8* in, int pixels)
...
+        out[2] = in[0];
+        out += 4; in += 2;
+    }
+}
+
+static void
+band316B(UINT8* out, const UINT8* in, int pixels)
+{
+    /* band 3 only, big endian */
+    int i;
```
Comment (radarhere) [inline src/libImaging/Unpack.c:1303]:
```suggestion
    int i;
    /* band 3 only, big endian */
```
Nitpick. Just to make it consistent with most of the other functions you've added here.

File: src/libImaging/Unpack.c [hunk 3]
```diff
@@ -1263,6 +1263,94 @@ band3I(UINT8* out, const UINT8* in, int pixels)
...
+        out[2] = in[1];
+        out += 4; in += 2;
+    }
+}
+
+static void
+band316L(UINT8* out, const UINT8* in, int pixels)
+{
+    /* band 3 only, little endian */
+    int i;
```
Comment (radarhere) [inline src/libImaging/Unpack.c:1347]:
```suggestion
    int i;
    /* band 3 only, little endian */
```

**Comment by wiredfool** [comment]
I want to review this before merge for safety.  Not that I’ll catch everything but I want to make a pass on it.

**Comment by kkopachev** [comment]
I've added tests for new unpackers. Travis failure is unrelated.

**Comment by kkopachev** [comment]
Any updates here? Would love this to make it into release.

**Comment by hugovk** [comment]
@wiredfool Do you think you'll have time to check this before release day?

**Comment by wiredfool** [comment]
Apparently not. It was on my list for last weekend, but I didn’t get to it.

**Comment by imfantuan** [comment]
Hey guys, any update here?

**Comment by aclark4life** [comment]
@imfantuan Needs a rebase at least

**Comment by kkopachev** [comment]
Rebased. 
I had to skip added tests for big endian system, following existing pattern.

**Comment by wiredfool** [comment]
This has major conflicts with a patch for a CVE that will be released in the next version.

**Comment by wiredfool** [comment]
See #5178

**Comment by hugovk** [comment]
Rebased in https://github.com/python-pillow/Pillow/pull/5178.

**Pull request closed without merging**